### DISCUSSION TITLE
5.4 when? *insert orangutan meme*

### DISCUSSION THREAD

**Discussion by deafstoryteller** [Sixze/ALS-Refactored#480] [Q&A] [answered]
All jokes aside, I'm curious if there's an estimated release window for official 5.4 support? 

Refactored is truly some of the most beautiful code I've seen, hats off to the original creator of ALS and even more so to Sixze and contributing devs for its transformation to the next level.

**Comment by DiamondGamesInc** [comment]
You guys are awesome, thank you so much for making this available, and doing all the heavy lifting! We know there's a ton of replication and issues with UE 5.3, so we'd love to put this directly into 5.4. Any idea when it'll be officially supported? Thank you again!

**Comment by MrWilsonWilson** [comment]
Not necessarily an answer but I can definitely jump on the appreciation wagon. It's very motivating to see the code being maintained and worked on, even more when we can see the changes following feedback (looking at that fov runtime change commit !)
All in all awesome, very grateful.

**Comment by Crazytalkbob** [comment]
Anyone run into issues with 5.4 yet? I just did a fresh install of 5.4 and the plugin and everything loaded.

**Reply by Jinkel92** [reply, marked as answer]
I already tested it and everything works fine.

**Reply by mustafakalaycii** [reply]
There is no problem in the editor when I move the project from version 5.3 to version 5.4, and I don't have any problems packaging the project, but when I open the project I have packaged, nothing about the Plugin works, how can I fix this problem.

**Comment by zSkeeter135** [comment]
From what i could see there is a couple of warnings because of deprecation. I noticed that the warnings happen in the AlsMacros with ALS_ENSURE and GET_VARARGS.  I removed the `FValidateArgsInternal(__VA_ARGS__);` line from ALS_ENSURE_IMPLEMENTATION macro in `AlsMacros.h` and changed `GET_VARARGS(FormattedMessage, FormattedMessageSize, FormattedMessageSize - 1, Format, Format);` to `GET_TYPED_VARARGS(TCHAR, FormattedMessage, FormattedMessageSize, FormattedMessageSize - 1, Format, Format);` in `bool UE_DEBUG_SECTION AlsEnsure::Execute` function in `AlsMacros.cpp`. From what I can tell everything is still working and no warnings for deprecation. I am very new when it comes to the system so I am sorry if this is wrong.

**Reply by Jinkel92** [reply]
For now it works because the obsolete stuff hasn't been removed yet. It may stop working in UE 5.5 or later.

**Reply by deafstoryteller** [reply]
Do you know how to remove the pop up 'plugin built for 5.3.2 attempt to load anyway?' and force 5.4.1? I changed the version via text edit of the ALS.uplugin file and still get the pop up but with 5.4.1

**Reply by Jinkel92** [reply]
No, I just accept. Meantime, we have to wait for updated version.

**Reply by Crazytalkbob** [reply]
Try setting it to 5.4 instead of 5.4.1

**Reply by onx2** [reply]
Updating `ALS.uplugin`'s config to `"EngineVersion": "5.4.0",` removes the pop up for me. No issues so far in 5.4.1 other than these deprecation warnings 😄

**Reply by watsonsong** [reply]
So is there some way to remove the `FValidateArgsInternal` warning?

**Reply by Thyke** [reply]
```
#pragma once

#include "Misc/AssertionMacros.h"

#define ALS_STRINGIFY_IMPLEMENTATION(Value) #Value

#define ALS_STRINGIFY(Value) ALS_STRINGIFY_IMPLEMENTATION(Value)

#define ALS_GET_TYPE_STRING(Type) \
    ((void)sizeof UEAsserts_Private::GetMemberNameCheckedJunk(static_cast<Type*>(nullptr)), TEXTVIEW(#Type))

#if DO_ENSURE && !USING_CODE_ANALYSIS

namespace AlsEnsure
{
    ALS_API bool UE_DEBUG_SECTION VARARGS Execute(std::atomic<bool>& bExecuted, bool bEnsureAlways, const ANSICHAR* Expression,
        const TCHAR* StaticMessage, const TCHAR* Format, ...);
}

#define ALS_ENSURE_IMPLEMENTATION(Capture, bEnsureAlways, Expression, Format, ...) \
    (LIKELY(Expression) || ([Capture]() UE_DEBUG_SECTION \
    { \
        static constexpr auto StaticMessage{TEXT("Ensure failed: " #Expression ", File: " __FILE__ ", Line: " ALS_STRINGIFY(__LINE__) ".")}; \
        static std::atomic<bool> bExecuted{false}; \
        \
        UE_VALIDATE_FORMAT_STRING(Format, ##__VA_ARGS__); \
        static constexpr ::UE::Assert::Private::FStaticEnsureRecord ENSURE_Static(Format, #Expression, __builtin_FILE(), __builtin_LINE(), bEnsureAlways); \
        if ((bEnsureAlways || !bExecuted.load(std::memory_order_relaxed)) && FPlatformMisc::IsEnsureAllowed() && ::UE::Assert::Private::EnsureFailed(bExecuted, &ENSURE_Static, ##__VA_ARGS__)) \
        { \
            PLATFORM_BREAK(); \
        } \
        \
        return false; \
    }()))

#define ALS_ENSURE(Expression) ALS_ENSURE_IMPLEMENTATION( , false, Expression, TEXT(""))
#define ALS_ENSURE_MESSAGE(Expression, Format, ...) ALS_ENSURE_IMPLEMENTATION(&, false, Expression, Format, ##__VA_ARGS__)
#define ALS_ENSURE_ALWAYS(Expression) ALS_ENSURE_IMPLEMENTATION( , true, Expression, TEXT(""))
#define ALS_ENSURE_ALWAYS_MESSAGE(Expression, Format, ...) ALS_ENSURE_IMPLEMENTATION(&, true, Expression, Format, ##__VA_ARGS__)

#else

#define ALS_ENSURE(Expression) (LIKELY(!!(Expression)))
#define ALS_ENSURE_MESSAGE(Expression, Format, ...) (LIKELY(!!(Expression)))
#define ALS_ENSURE_ALWAYS(Expression) (LIKELY(!!(Expression)))
#define ALS_ENSURE_ALWAYS_MESSAGE(Expression, Format, ...) (LIKELY(!!(Expression)))

#endif
```


```
#include "Utility/AlsMacros.h"
#include "CoreGlobals.h"
#include "Templates/Function.h"
#include "Engine/EngineTypes.h" // FOverlapResult ve FMTDResult
#include "DrawDebugHelpers.h"   // ENABLE_DRAW_DEBUG ve DrawDebugLine
#include "Components/PrimitiveComponent.h" // FBodyInstance

#if DO_ENSURE && !USING_CODE_ANALYSIS

bool UE_DEBUG_SECTION AlsEnsure::Execute(std::atomic<bool>& bExecuted, const bool bEnsureAlways, const ANSICHAR* Expression,
    const TCHAR* StaticMessage, const TCHAR* Format, ...)
{
    if ((bExecuted.load(std::memory_order_relaxed) && !bEnsureAlways) || !FPlatformMisc::IsEnsureAllowed())
    {
        return false;
    }

    bExecuted.store(true, std::memory_order_relaxed);

    static constexpr auto FormattedMessageSize{ 4096 };
    TCHAR FormattedMessage[FormattedMessageSize];

    va_list Args;
    va_start(Args, Format);
    TCString<TCHAR>::GetVarArgs(FormattedMessage, FormattedMessageSize, Format, Args);
    va_end(Args);

    if (UNLIKELY(GetEnsureHandler() && GetEnsureHandler()({ Expression, FormattedMessage })))
    {
        return false;
    }

    UE_LOG(LogOutputDevice, Warning, TEXT("%s"), StaticMessage);
    UE_LOG(LogOutputDevice, Warning, TEXT("%s"), FormattedMessage);

    PrintScriptCallstack();

    if (!FPlatformMisc::IsDebuggerPresent())
    {
        FPlatformMisc::PromptForRemoteDebugging(true);
        return false;
    }

#if UE_BUILD_SHIPPING
    return true;
#else
    return !GIgnoreDebugger;
#endif
}

#endif
```

Old API (GET_VARARGS): This macro was used for handling various variable arguments (varargs). However, it has been deprecated and a new macro is recommended.
New API (GET_TYPED_VARARGS): The new macro is used for handling variable arguments based on a specific character type (CharType). This provides a more type-safe approach.
Performed Update: We replaced the GET_VARARGS macro with the GET_TYPED_VARARGS macro. This ensures compatibility with the new API and removes deprecation warnings.


**Changes:**
Old version:
`GET_VARARGS(FormattedMessage, FormattedMessageSize, FormattedMessageSize - 1, Format, Format);`

New version:
```
va_list Args;
va_start(Args, Format);
TCString<TCHAR>::GetVarArgs(FormattedMessage, FormattedMessageSize, Format, Args);
va_end(Args);
```


and also into CameraComponent.Cpp

`#include "Engine/OverlapResult.h"`

Don't forget to add. Do not accept the codes as correct. The repo owner will arrange it best.

**Comment by gustavorios2929** [comment]
I'm waiting too ❤️
.
.
.
I don't like warnings 😅

**Comment by MrWilsonWilson** [comment]
When you guys say "no issues" do you mean it compiles and packages ? I can compile it fine but package throws me some errors

**Reply by MrWilsonWilson** [reply]
Update : setting engine version to the right one indeed fixes all problems

**Reply by mustafakalaycii** [reply]
There is no problem in the editor when I move the project from version 5.3 to version 5.4, and I don't have any problems packaging the project, but when I open the project I have packaged, nothing about the Plugin works, how can I fix this problem.

**Comment by sandiz** [comment]
are you guys seeing issue with the feet looking weird in edit mode/thumbnail -- it fixes itself in gameplay so not a blocker

<img width="1117" alt="image" src="https://github.com/Sixze/ALS-Refactored/assets/5438703/e041dfa9-a00e-4eb3-b07b-62e01a8ac7c6">

**Reply by Magic-Summer** [reply]
Yes, I see this too. Must be from a recent update. As you say, in PIE it disappears.

**Comment by GStreetGames** [comment]
Rather than asking when, I want to ask what is left that's holding back a new release? Is it the 10 issues, or is there something more? I'd love to dig in and help get this going, but have no idea what to look into. It's obvious that more talent is needed to help this along, so we should all be asking this question.
